src/main/java/org/symphonyoss/symphony/messageml/elements/Tag.java

```diff
@@ -0,0 +1,268 @@
+  }
+
+  @Override
+  public ObjectNode asEntityJson(ObjectNode parent) {
+    ObjectNode node = super.asEntityJson(parent);
+    if (instrument != null) {
+      enhanceInstrumentEntityJson(node);
+    }
+    return node;
+  }
+
+
+  private void enhanceInstrumentEntityJson(ObjectNode node) {
+    ArrayNode idArray = ArrayNode.class.cast(node.get(ID_FIELD));
+    idArray.add(
+        buildNode("org.symphonyoss.fin.security.id.uniqueId", instrument.getUniqueId()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.id.fullBbgTicker",
+        instrument.getFullBbgCompTicker()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.bbgcompticker",
+        instrument.getBbgCompTicker()));
+    String isin = null;
+    if (instrument instanceof Index) {
+      isin = Index.class.cast(instrument).getIsin();
+    } else if (instrument instanceof Equity) {isin = Equity.class.cast(instrument).getIsin();}
+    idArray.add(buildNode("org.symphonyoss.fin.security.id.isin", isin));
+    idArray.add(buildNode("org.symphonyoss.fin.security.id.figi", instrument.getFigi()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.id.figiTicker",
+        instrument.getFigiTicker()));
+    idArray.add(
+        buildNode("org.symphonyoss.fin.security.id.lei", instrument.getFigiTicker()));
+    idArray.add(
+        buildNode("org.symphonyoss.fin.security.id.localCode", instrument.getLocalCode()));
+    String operationalMic = null;
+    if (instrument instanceof Equity) {
+      operationalMic = Equity.class.cast(instrument).getOperationalMic();
+    }
+    idArray.add(buildNode("org.symphonyoss.fin.security.id.operationalMic",
+        operationalMic));
+    String countryCode = null;
+    if (instrument instanceof Equity) {
+      countryCode = Equity.class.cast(instrument).getCountryCode();
+    } else if (instrument instanceof Index) {
+      countryCode = Index.class.cast(instrument).getCountryCode();
+    }
+    idArray.add(buildNode("org.symphonyoss.fin.security.countryCode", countryCode));
+    String countryName = null;
+    if (instrument instanceof Equity) {
+      countryName = Equity.class.cast(instrument).getCountryName();
+    }
+    idArray.add(buildNode("org.symphonyoss.fin.security.countryName", countryName));
+    String exchangeName = null;
+    if (instrument instanceof Equity) {
+      exchangeName = Equity.class.cast(instrument).getExchangeName();
+    }
+    idArray.add(buildNode("org.symphonyoss.fin.security.exchangeName", exchangeName));
+    idArray.add(buildNode("org.symphonyoss.fin.security.displayName", instrument.getDisplayName()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.currency", instrument.getCurrency()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.instrumentTypeCode",
+        instrument.getInstrumentTypeCode()));
+    idArray.add(buildNode("org.symphonyoss.fin.security.instrumentTypeName",
+        instrument.getInstrumentTypeName()));
+  }
```

Comment:
Would it be easier if we collapsed all these types (Equity, FxCross, Index..) into one and just avoid building nodes for null values? it seems that this code has a lot of instance of and cast that might be unnecessary. Another possibility would be to move up all parameters  which we are casting to the Instrument object. This code is pretty complexe and I am not sure we need this much complexity here